### ISSUE TITLE
Importing schemas from kafka after container restart crashes on "SELECT nextval('globalidsequence')"

### ISSUE THREAD

**Issue by petolexa** [Apicurio/apicurio-registry#1500]
Hello,

we have an issue with re-importing messages from Kafka topic back to internal h2 DB, that results into container crash. Here are the steps we followed:

1. Exporting data with [exporter](https://github.com/Apicurio/apicurio-registry/tree/master/utils/exportV1) 
2. Importing it to apicurio-registry-kafkasql:2.0.0.Final container connected to empty Kafka topic
3. Everything works fine until we re-deploy the Kubernetes deployment, whitch normally results to pods restart.
4. After restart, the pod will never run successfully again, it keeps crashing in loop.

All imports from Kafka after restart were successful, so it was not very visible in regitstry log, but when I increased DB trace level:
`-D%prod.quarkus.datasource.jdbc.url=jdbc:h2:mem:registry_db;TRACE_LEVEL_SYSTEM_OUT=2`

after last import I've got this:
```
2021-05-15 14:05:41,632 DEBUG [io.api.reg.sto.imp.kaf.sql.KafkaSqlSink_Subclass] (KSQL Kafka Consumer Thread) Kafka message successfully processed. Notifying listeners of response.
2021-05-15 14:05:41,632 DEBUG [io.api.reg.sto.imp.kaf.sql.KafkaSqlSink_Subclass] (KSQL Kafka Consumer Thread) Processing Kafka message with UUID: ba36c16c-c4de-45b7-bea7-340df895cca1
2021-05-15 14:05:41,633 TRACE [org.jdb.v3.cor.Jdbi] (KSQL Kafka Consumer Thread) Jdbi [org.jdbi.v3.core.Jdbi@2b186595] obtain handle [org.jdbi.v3.core.Handle@cfc62d2] in 0ms
2021-05-15 14:05:41 jdbc[3]: 
/*SQL #:1*/SELECT MAX(globalId) FROM versions;
2021-05-15 14:05:41,634 INFO  [io.api.reg.sto.imp.sql.AbstractSqlRegistryStorage] (KSQL Kafka Consumer Thread) Resetting globalId sequence to 12517376
2021-05-15 14:05:41 jdbc[3]: 
/*SQL #:1 t:3*/SELECT nextval('globalidsequence');
2021-05-15 14:05:41 jdbc[3]: 
/*SQL #:1*/SELECT nextval('globalidsequence');
2021-05-15 14:05:41 jdbc[3]: 
/*SQL #:1*/SELECT nextval('globalidsequence');

... thousands of these lines ...

2021-05-15 14:05:46 jdbc[3]: 
/*SQL #:1*/SELECT nextval('globalidsequence');
2021-05-15 14:05:46 jdbc[3]: 
/*SQL #:1*/SELECT nextval('globalidsequence');
2021-05-15 14:05:46 jdbc[3]: 
/*SQL #:1*/SELECT nextval('globalidsequence');
```

then
```
2021-05-15 14:11:17 database: closing mem:registry_db from shutdown hook
2021-05-15 14:11:17 database: disconnecting session #3
2021-05-15 14:11:17 database: disconnected session #3
2021-05-15 14:11:17 database: disconnecting session #4
...
2021-05-15 14:11:17 database: disconnecting session #22
2021-05-15 14:11:17 database: disconnected session #22
```

and this at the end:
```
2021-05-15 14:11:17 database: closing mem:registry_db
2021-05-15 14:11:17 database: closed
2021-05-15 14:11:17 jdbc[3]: exception
2021-05-15 14:11:17,472 TRACE [org.jdb.v3.cor.Handle] (KSQL Kafka Consumer Thread) Handle [org.jdbi.v3.core.Handle@35dc7564] released
2021-05-15 14:11:17 jdbc[3]: exception
2021-05-15 14:11:17 jdbc[3]: exception
2021-05-15 14:11:17,491 DEBUG [io.api.reg.sto.imp.kaf.sql.KafkaSqlSink_Subclass] (KSQL Kafka Consumer Thread) Registry exception detected: org.jdbi.v3.core.result.ResultSetException: Could not get result set [statement:"SELECT nextval('globalidsequence')", arguments:{positional:{}, named:{}, finder:[]}]
2021-05-15 14:11:17,491 DEBUG [io.api.reg.sto.imp.kaf.sql.KafkaSqlSink_Subclass] (KSQL Kafka Consumer Thread) Processing Kafka message with UUID: c363bc84-dbd6-4546-a998-a427c0d3e2d3
2021-05-15 14:11:17,491 DEBUG [io.api.reg.sto.imp.kaf.sql.KafkaSqlSink_Subclass] (KSQL Kafka Consumer Thread) Unexpected exception detected: Error injecting javax.transaction.TransactionManager io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.transactionManager
```

I think the issue is connected with with the high globalId sequence number:
`2021-05-15 14:05:41,634 INFO  [io.api.reg.sto.imp.sql.AbstractSqlRegistryStorage] (KSQL Kafka Consumer Thread) Resetting globalId sequence to 12517376`

as our glogalIds of imported schemas are following:
```
1
65536
458753
...
1769472
1769473
...
4259841
4325377
4456448
4456449
4915201
4980736
...
6815745
6881281
6946817
7012353
7077889
...
12320768
12517376
```

Is it tpossible, that the import api can handle these globalIds (as it is first time imported successfully), but import after container restart is different approach and cannot handle if global IDs are not in sequence?
Could you please suggest how can we handle this in version 2.0.0.Final?

**Comment by petolexa** [comment]
I found a workaround, that seems to be working (pods start normally after restart, no deadloop in DB):

I exchanged retrieving old globalIds in exporter [from metadata](https://github.com/Apicurio/apicurio-registry/blob/c228b851f39a69eb27b07b834ac132a59508644c/utils/exportV1/src/main/java/io/apicurio/registry/utils/export/Export.java#L161) for a sequence of globalIds starting from 1 and increased by 1:

```
AtomicInteger globalIdSeq = new AtomicInteger(1);

...

versionEntity.globalId = globalIdSeq.getAndIncrement();
```

But it is not very robust, as you can only use it for empty h2 DB (and empty kafka topic ideally). Plus I guess retrieving globalId from metadata has its reasons.
Do you think we can use it as one-time import workaround?

I tried importing new schema then via REST api and it was registrered successfully with globalId increased by 1 comparing to to the previous one.

**Comment by EricWittmann** [comment]
This is a fine workaround if you don't care about preserving globalIds.  However, if you have existing Kafka apps with messages serialized using our Serdes, then you'll likely have the globalId embedded in the message payload.  So if you restore without preserving globalIds, then deserialization of those old messages will break.

If that's not an issue, then your workaround is fine.

I don't think the problem is likely related to the lack of proper sequencing.  Instead, I suspect that the issue is with the max size of your globalIds.  I suspect the problem is here:

https://github.com/Apicurio/apicurio-registry/blob/master/app/src/main/java/io/apicurio/registry/storage/impl/sql/AbstractSqlRegistryStorage.java#L2262-L2275

Originally when I wrote that code I tried to update the sequence value using SQL specific to each DB we support, but I could never get it to work.  So I fell back to just incrementing the sequence in a loop.  If you have a very large globalId, it's possible this is a horrible idea.

**Comment by EricWittmann** [comment]
@famartinrh ^^^

**Comment by petolexa** [comment]
Thank you for the answer and for suggestions. I checked with my colleagues and you are right, we cannot use 'simple' sequence workaround as globalIds have to stay preserved.

Prevoiusly we used Streams version and I think our large global IDs are connected with [this code](https://github.com/Apicurio/apicurio-registry/blob/e3f786ab30af5412d853a4acde793c6e09584724/storage/streams/src/main/java/io/apicurio/registry/streams/StreamsPropertiesImpl.java#L21-L23):
```
public long toGlobalId(long offset, int partition) {
       return getBaseOffset() + (offset &lt;&lt; 16) + partition;      
}
```
Nevertheless, globalIds are pretty big, but globalidsequence and globalId column in h2 database are 'BIGINT' and globalId in java is 'long', so we are safely far from maximum.

In the log, I see this line from the code you suggested:
`"Resetting globalId sequence to {}", id);`
but I do not get to this:
`"Successfully reset globalId to {}", nextGlobalId(handle);`
and after loop if sequences I end up there:
```
2021-05-15 14:05:46 jdbc[3]: 
/*SQL #:1*/SELECT nextval('globalidsequence');
2021-05-15 14:11:17 database: closing mem:registry_db from shutdown hook
2021-05-15 14:11:17 database: disconnecting session #3
```

So basically what I need to do is to give enough time to to your loop to iterate up to my globalId, right? Is there a way how to do it? Can I make DB session open longer?

Speaking about moving from Streams and mentioned issues, I alho have 2 side-questions:
- Is tere a persistence type you prefer/recommend between Kafkasql and PostgreSQL?
- Could you please estimate when new Final vrsion will be released? There are some useful bugfixes in lates-snapshot version in comparison with v2.0.0.Final.

**Comment by petolexa** [comment]
Thank you for the recommendation, we were discussing if it is worth trying PostgreSQL instad of Kafka in v2, but the biggest concern is about the performace - PostrgreSQL should be significantly slower as Kafkasql version uses in-memory DB. On the other hand, importing messages from Kafka topic on every re-deployment migth bring issues like thisone.

We have 53 artifacts now, at the time of writing it was 51 artifact = 67 unique items (some artifacts with more versions). So nothing big :) But the number range of globalIds is big.

Regarding giving time to loop, I tried adding this to JAVA_OPTIONS:
`-D%prod.quarkus.datasource.jdbc.url=jdbc:h2:mem:registry_db;DB_CLOSE_DELAY=60`
but a) it does not work so far and b) I do not know if I go right way with giving the loop enogh time.

**Comment by petolexa** [comment]
I've tried the fix above with the latest-snapshot container and it works nicely for me:
```
2021-05-20 14:28:47 jdbc[3]: 
/*SQL #:1*/SELECT MAX(globalId) FROM versions;
2021-05-140 14:28:47 INFO <> [io.apicurio.registry.storage.impl.sql.AbstractSqlRegistryStorage] (KSQL Kafka Consumer Thread) Resetting globalId sequence
2021-05-20 14:28:47 jdbc[3]: 
/*SQL l:46 t:2*/ALTER SEQUENCE globalidsequence RESTART WITH ? {1: 13893633};
2021-05-140 14:28:47 INFO <> [io.apicurio.registry.storage.impl.sql.AbstractSqlRegistryStorage] (KSQL Kafka Consumer Thread) Successfully reset globalId to 13893633
2021-05-140 14:28:47 TRACE <> [org.jdbi.v3.core.Handle] (KSQL Kafka Consumer Thread) Handle [org.jdbi.v3.core.Handle@468aeaf2] released
```

no container crash, no issues after container restart and re-import from Kafka topic. Thank you very much for quick fix.

Can I close this? Or maybe wait a while maybe, if your colleagues could estimate next minor Final release date?

**Comment by EricWittmann** [comment]
( closing this as fixed by https://github.com/Apicurio/apicurio-registry/pull/1507 )
